### DISCUSSION TITLE
Multiple Cubes with Mesh Generation

### DISCUSSION THREAD

**Discussion by AryGP** [idaholab/moose#28090] [Q&A Meshing]
Hello,
I am trying to simulate the thermal expansion of two copper wafers using MOOSE and visualizing on Paraview. I am struggling to create 2 objects in moose. I am using the solid_mechanics/thermal_expansion/constant_expansion_stress_free_temp.i
This is what I have, I just want to make 2 cubes and see both of them expand and come in contact. 
```
[Mesh]
  type = GeneratedMesh
  dim = 3
  nx = 2
  ny = 2
  nz = 2
[]

[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
[]

[AuxVariables]
  [./temp]
  [../]
[]

[Functions]
  [./temperature_load]
    type = ParsedFunction
    expression = t*(500.0)+300.0
  [../]
[]

[Physics]
  [SolidMechanics]
    [QuasiStatic]
      [./all]
        strain = SMALL
        incremental = true
        add_variables = true
        eigenstrain_names = eigenstrain
        generate_output = 'strain_xx strain_yy strain_zz'
      [../]
    [../]
  [../]
[]

[AuxKernels]
  [./tempfuncaux]
    type = FunctionAux
    variable = temp
    function = temperature_load
  [../]
[]

[BCs]
  [./x_bot]
    type = DirichletBC
    variable = disp_x
    boundary = left
    value = 0.0
  [../]
  [./y_bot]
    type = DirichletBC
    variable = disp_y
    boundary = bottom
    value = 0.0
  [../]
  [./z_bot]
    type = DirichletBC
    variable = disp_z
    boundary = back
    value = 0.0
  [../]
[]

[Materials]
  [./elasticity_tensor]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 2.1e5
    poissons_ratio = 0.3
  [../]
  [./small_stress]
    type = ComputeFiniteStrainElasticStress
  [../]
  [./thermal_expansion_strain]
    type = ComputeThermalExpansionEigenstrain
    stress_free_temperature = 298
    thermal_expansion_coeff = 1.3e-5
    temperature = temp
    eigenstrain_name = eigenstrain
  [../]
[]

[Executioner]
  type = Transient
  solve_type = 'PJFNK'

  l_max_its = 50
  nl_max_its = 50
  nl_rel_tol = 1e-12
  nl_abs_tol = 1e-10
  l_tol = 1e-9

  start_time = 0.0
  end_time = 0.075
  dt = 0.0125
  dtmin = 0.0001
[]

[Outputs]
  exodus = true
  checkpoint = true
[]

[Postprocessors]
  [./strain_xx]
    type = ElementAverageValue
    variable = strain_xx
  [../]
  [./strain_yy]
    type = ElementAverageValue
    variable = strain_yy
  [../]
  [./strain_zz]
    type = ElementAverageValue
    variable = strain_zz
  [../]
  [./temperature]
    type = AverageNodalVariableValue
    variable = temp
  [../]
[]
```

**Comment by GiudGiud** [comment]
Hello

Try this to make two cubes
```
[Mesh]
  [cube1]
    type = GeneratedMeshGenerator
    dim = 3
    nx = 2
    ny = 2
    nz = 2
    subdomain_name = 'cube1'
  []
  [cube2]
    type = GeneratedMeshGenerator
    dim = 3
    xmin = 1
    xmax = 2
    nx = 2
    ny = 2
    nz = 2
    subdomain_ids = '1'
    subdomain_name = 'cube2'
  []
  [stitch]
    type = StitchedMeshGenerator
    inputs = 'cube1 cube2'
    stitch_boundaries_pairs = 'right left'
[]
[]
```

The two meshes will have different subdomains so you can define different materials on them for example

**Reply by AryGP** [reply]
Awesome, that worked. Thanks!
The only issue I'm having now is that whenever I take the file to paraview (visualizer for moose), it's not able to create the cubes as it was able to before with just the 1. 
<img width="579" alt="image" src="https://github.com/user-attachments/assets/4c3f8cf3-08d1-4117-8807-b26aac0e939c">

**Reply by AryGP** [reply]
also, after the 3rd time step, in the ran test, 
     30 Linear |R| = 6.013082e-14
this is red, while all the other are green.
<img width="301" alt="image" src="https://github.com/user-attachments/assets/37010e61-504b-4fde-b39e-0bcf1ec97eaa">

**Reply by GiudGiud** [reply]
the 30th is red because the residual increased instead of decreased. this is not cause for concern

**Reply by AryGP** [reply]
<img width="525" alt="image" src="https://github.com/user-attachments/assets/1c872c4d-73d5-495b-a5da-353fb11e45b9">

**Reply by AryGP** [reply]
<img width="221" alt="image" src="https://github.com/user-attachments/assets/3bef36cb-279f-4c44-b289-916a8939b063">

**Reply by GiudGiud** [reply]
in StitchedMeshGenerator set `prevent_boundary_ids_overlap = False`

**Reply by AryGP** [reply]
Wow, thanks. That worked. Do you mind explaining the syntax of the coordinate system? I am now trying to create a small gap between those two cubes instead of having them touch so I can see at what distance I can set them apart so they come in contact under increase in temp.

**Reply by GiudGiud** [reply]
if the cubes should be separate do not stitch them 

you can use a CombinerGenerator to just have the two blocks sitting next to each other

**Reply by AryGP** [reply]
Ok.
1. Also, on a separate note. If I want to change the material of the cube, how would I do that? Is that something I need to do in moose or in paraview?
2. Is it possible to have a cube and a cylinder together as well instead of 2 cubes?

**Reply by GiudGiud** [reply]
these are both trivial things to do in moose
please follow the training material here
https://mooseframework.inl.gov/getting_started/examples_and_tutorials/index.html

**Reply by AryGP** [reply]
I tried this
  []
  [cylinder]
    type = GeneratedMeshGenerator
    dim = 3
    shape = 'cylinder'
    radius = 1.0
    height = 2.0
    num_rings = 2
    num_segments = 8
    num_axial = 2
    subdomain_name = 'cylinder'
  []
  
  but it seems like none of these are able to be recognized in moose

**Reply by GiudGiud** [reply]
```
shape = 'cylinder'
```
this is not a parameter

**Reply by AryGP** [reply]
ok. 
[Mesh]
  [cube1]
    type = GeneratedMeshGenerator
    dim = 3
    nx = 2
    ny = 2
    nz = 2
    subdomain_name = 'cube1'
  []
  [cylinder]
    type = GeneratedMeshGenerator
    dim = 3
    radius = 1.0
    height = 2.0
    num_rings = 2
    num_segments = 8
    num_axial = 2
    subdomain_name = 'cylinder'
  []
  [stitch]
    type = StitchedMeshGenerator
    prevent_boundary_ids_overlap = False
    inputs = 'cube1 cylinder'
    stitch_boundaries_pairs = 'right left'
[]
[]

It's still saying I have unused parameters

**Reply by AryGP** [reply]
None of the examples MOOSE provides has a solid cylinder mesh

**Reply by GiudGiud** [reply]
use these two to build a cylinder from a circle
https://mooseframework.inl.gov/source/meshgenerators/ConcentricCircleMeshGenerator.html
https://mooseframework.inl.gov/source/meshgenerators/AdvancedExtruderGenerator.html

**Comment by AryGP** [comment]
I was able to create the Concentric circle but I can't get the extrude generator to work. 
[Mesh]
  [cube1]
    type = GeneratedMeshGenerator
    dim = 3
    nx = 2
    ny = 2
    nz = 2
    subdomain_name = 'cube1'
  []
  [circle]
    type = ConcentricCircleMeshGenerator
    num_sectors = 6
    radii = '0.2546 0.3368 0.3600 0.3818 0.3923 0.4025 0.4110 0.4750'
    rings = '5 3 2 1 1 1 1 3 5'
    has_outer_square = on
    pitch = 1.42063
    #portion = left_half
    preserve_volumes = off
    smoothing_max_it = 3
  []
  [extrude]
    type = AdvancedExtruderGenerator
    input = cylinder
    heights = '1 2 3'
    num_layers = '1 2 3'
    direction = '0 0 1'
    bottom_sideset = '4'
    top_sideset = '5'
    subdomain_swaps = '0 1;
                       0 2;
                       0 3'
  []
  [stitch]
    type = StitchedMeshGenerator
    prevent_boundary_ids_overlap = False
    inputs = 'cube1 cylinder'
    stitch_boundaries_pairs = 'right left'
[]
[]


*** ERROR ***
Your MeshGenerator tree contains multiple possible generator outputs :
"stitch and one or more of the following from an independent set: "extrude"

This may be due to a missing dependency or may be intentional. Please either
- check that all the mesh generators are connected as a tree and culminate in a single final mesh. Having one wrong 'input=mg' parameter is the most common error
- add additional dependencies to remove the ambiguity if you are using a user-built MeshGenerator
- if you intend to execute a subset of the defined generators (uncommon), select the final MeshGenerator in the [Mesh] block with the "final_generator" parameter.

**Reply by GiudGiud** [reply]
the error message explains what is happening.
you need every mesh generator to form a connected graph with a single output mesh (the tree)
